File: Leetcode_cpp_py/3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 * 
 * 滑窗法，维护一个map，key为字符，value为key在串中的位置
 * 1. 记录子串开始的位置start，
 * 2. 遇到重复的字符则后移一位start+1
 * 3. 不重复且子串最大长度增加，则更新maxLength
 */
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<int, int> mymap;
        int start, maxLength;
        start = maxLength = 0;

        for(int i = 0; i < s.length(); i++){
            if(mymap.count(s[i]) && start <= mymap[s[i]]){
                start = mymap[s[i]] + 1;
            }else{
                maxLength = max(maxLength, i - start + 1);
            }
            mymap[s[i]] = i;
        }
        return maxLength;
    }
};

File: Leetcode_cpp_py/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start = maxLength = 0
        mymap = {}

        for i in range(len(s)):
            if s[i] in mymap and start <= mymap[s[i]]:
                start = mymap[s[i]] + 1 #有重复的，更新起点
            else:
                maxLength = max(maxLength, i - start + 1)   #取较大一个尾部
            
            mymap[s[i]] = i #插入新键值

        return maxLength

